Handle non-string values in WP_HTML_Tag_Processor

kraftbj opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Adds handling of non-string values passed to the WP_HTML_Tag_Processor by throwing a _doing_it_wrong and setting it to an empty string to avoid further issues.

Trac ticket: https://core.trac.wordpress.org/ticket/63854

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

kraftbj committed:
Handle non-string values in WP_HTML_Tag_Processor

sirreal commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -72,6 +72,25 @@ public function test_has_self_closing_flag_matches_input_html( $html, $flag_is_s
 	}
 
+	/**
+	 * @ticket 6.9.0
```
Should be the ticket number:
```suggestion
	 * @ticket 63854
```

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -830,10 +830,19 @@ class WP_HTML_Tag_Processor {
 	 * Constructor.
 	 *
 	 * @since 6.2.0
+	 * @since 6.9.0 Executes doing_it_wrong() when the HTML parameter is not a string.
```
I don't know if there are broad guidelines on when to include since tags. To me, this doesn't seem significant enough to note in the function docblock. I'd exclude it.

kraftbj replied:
No objection from me to remove it. I could be convinced either way; erring on more rather than less. Removed in e366ee3

kraftbj committed:
Trac ticket number

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

kraftbj committed:
Remove the since annotation

sirreal commented:
I think ‎`WP_HTML_Processor::create_fragment()` and `WP_HTML_Processor::create_full_parser()` are also susceptible. Those can also `_doing_it_wrong()`, but they're simpler in that they can return null. Do you mind adding those changes here?

sirreal commented:
@kraftbj Just checking in, I'd love to get this in 6.9. Are you able to get back to it or would you like help to finish up?

kraftbj commented:
Apologies—death in the family after the last comment and it fell off my radar. I'll refresh the branch and make the updates to the other functions now.

kraftbj committed:
Merge remote-tracking branch 'origin/trunk' into update/html-tag-proecssor-null

kraftbj committed:
Defensive coding for non-strings in ::create_fragment and ::create_full_parser

westonruter commented on tests/phpunit/tests/html-api/wpHtmlTagProcessor.php:
```diff
@@ -72,6 +72,25 @@ public function test_has_self_closing_flag_matches_input_html( $html, $flag_is_s
+	 *
+	 * @covers WP_HTML_Tag_Processor::__construct
+	 */
+	public function test_constructor_validates_html_parameter() {
```
You'll add similar tests for `WP_HTML_Processor`?

westonruter reviewed (approved):
Looks good!

sirreal committed:
Add HTML processor doing it wrong tests

sirreal committed:
Use expectedIncorrectUsage test annotation over method

This makes it easier to run the HTML API tests in isolation without
the entire phpunit configuration.

sirreal reviewed (approved):
Thank you! I added tests for the HTML processor.

This is ready to go 👍

sirreal committed:
Merge branch 'trunk' into update/html-tag-proecssor-null
